src/jupyterhub_config/custom_docker_spawner.py

```diff
@@ -229,24 +231,42 @@ def _configure_environment(self, user_dir: Path, user_env_dir: Path, username: s
 
         self.environment["SHELL"] = "/usr/bin/bash"
 
-        if self._is_rw_minio_user():
-            self.log.info(
-                f"MinIO read/write user detected: {self.user.name}. Setting up minio_rw credentials."
+        # Get user-specific MinIO credentials from governance service
+        # Check if KBASE_AUTH_TOKEN is available (should be set by pre_spawn_start)
+        kbase_token = self.environment.get("KBASE_AUTH_TOKEN")
+        if not kbase_token:
+            raise ValueError(
+                "KBASE_AUTH_TOKEN not found in spawner environment. Authentication required for governance client."
             )
-            self.environment["MINIO_ACCESS_KEY"] = self.environment[
-                "MINIO_RW_ACCESS_KEY"
-            ]
-            self.environment["MINIO_SECRET_KEY"] = self.environment[
-                "MINIO_RW_SECRET_KEY"
-            ]
-            # USAGE_MODE is used by the setup.sh script to determine the appropriate configuration for the user.
-            self.environment["USAGE_MODE"] = "dev"
+        client = DataGovernanceClient(kbase_token=kbase_token)
+
+        # Get credentials for the current user - it will also create a new user if it doesn't exist
+        credentials = client.get_credentials()
+        self.log.info(
+            f"Retrieved governance credentials for user {username}: access_key='{credentials.access_key}'"
+        )
+
+        if self.environment.get("USE_DATA_GOVERNANCE_CREDENTIALS", "false") == "true":
+            # Set user-specific MinIO credentials
+            self.environment["MINIO_ACCESS_KEY"] = credentials.access_key
+            self.environment["MINIO_SECRET_KEY"] = credentials.secret_key
```

Comment:
If a user accidentally leaks their creds, is there a way to remove their old creds and get a new set?

Reply:
We do not store users' MinIO credentials. Instead, new credentials are generated each time a user logs in. Users can also manually rotate their credentials by calling `get_minio_credentials`, which is the same method invoked during the login process.

Reply:
:+1:

Reply:
Saves us a lot of maintain and development cycle. I didn't expect users will need to login to minIO every time. It's also in the [design doc](https://github.com/kbase/cdm-minio-service/blob/main/docs/design.md) - no one ever raised any concern.

Reply:
Looks like no one ever reviewed it :/